components/psst/browser/core/psst_rule.cc

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/psst/browser/core/psst_rule.h"
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "base/files/file_path.h"
+#include "base/json/json_reader.h"
+#include "base/types/expected.h"
+#include "extensions/common/url_pattern.h"
+#include "net/base/url_util.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+#include "url/gurl.h"
+#include "url/origin.h"
+#include "url/url_constants.h"
+
+namespace {
+
+// Format:
+// [
+//   {
+//     "include": [
+//       "https://twitter.com/*"
+//     ],
+//     "exclude": [
+//     ],
+//     "version": 1,
+//     "test_script": "twitter/test.js",
+//     "policy_script": "twitter/policy.js"
+//   }, ...
+// ]
+// Note that "test_script" and "policy_script" give paths
+// relative to the component under scripts/
+
+// psst.json keys
+const base::FilePath::CharType kInclude[] = FILE_PATH_LITERAL("include");
+const base::FilePath::CharType kExclude[] = FILE_PATH_LITERAL("exclude");
+const base::FilePath::CharType kTestScript[] = FILE_PATH_LITERAL("test_script");
+const base::FilePath::CharType kPolicyScript[] =
+    FILE_PATH_LITERAL("policy_script");
+const base::FilePath::CharType kVersion[] = FILE_PATH_LITERAL("version");
+
+bool GetURLPatternSetFromValue(const base::Value* value,
+                               extensions::URLPatternSet* result) {
+  if (!value->is_list()) {
+    return false;
+  }
+  std::string error;
+  bool valid = result->Populate(value->GetList(), URLPattern::SCHEME_HTTPS,
+                                false, &error);
+  if (!valid) {
+    DVLOG(1) << error;
+  }
+  return valid;
+}
+
+bool GetFilePathFromValue(const base::Value* value, base::FilePath* result) {
+  if (!value->is_string()) {
+    return false;
+  }
+  *result = base::FilePath(value->GetString());
+  return true;
+}
+
+}  // namespace
+
+namespace psst {
+
+PsstRule::PsstRule() = default;
+PsstRule::~PsstRule() = default;
+PsstRule::PsstRule(const PsstRule& other) {
+  include_pattern_set_ = other.include_pattern_set_.Clone();
+  exclude_pattern_set_ = other.exclude_pattern_set_.Clone();
+  test_script_path_ = other.test_script_path_;
+  policy_script_path_ = other.policy_script_path_;
+  version_ = other.version_;
+}
+
+// static
+void PsstRule::RegisterJSONConverter(
```

Comment:
Would you be able to use an `json_schema_compiler` IDL here? I would rather we avoid using `JSONValueConverter`, as has been also the case upstream.

Reply:
moved to follow-up